### PR TITLE
[LSM] Scaffolding

### PR DESCRIPTION
**Pull request by sampocs**
Closes: #XXX

## Author's Checklist

I have...

- [ ] Run and PASSED locally all GAIA integration tests
- [ ] If the change is contentful, I either:
    - [ ] Added a new unit test OR 
    - [ ] Added test cases to existing unit tests
- [ ] OR this change is a trivial rework / code cleanup without any test coverage

If skipped any of the tests above, explain.

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] manually tested (if applicable)
- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] README.md 
      - [ ] not documented

### CODE REVIEW

File: proto/stride/stakeibc/lsm.proto [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+message LSMTokenDeposit {
+  string denom = 1;
+  string chain_id = 2;
+  string validator_address = 3;
+  string amount = 4 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable) = false
+  ];
+  uint64 record_id = 5;
```
Comment (ethan-stride) [inline proto/stride/stakeibc/lsm.proto:27]: If the denom is going to contain the unique ending to the NFT like "cosmosval42" why do we also need this field for the record_id "42".  Won't we always want to look up using the full denom and this would create additional logic to go back and forth from the denom to the record_id alone?

File: proto/stride/stakeibc/lsm.proto [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import "cosmos/base/v1beta1/coin.proto";
+import "gogoproto/gogo.proto";
+
+option go_package = "github.com/Stride-Labs/stride/v8/x/stakeibc/types";
+
+enum LSMDepositStatus {
+  option (gogoproto.goproto_enum_prefix) = false;
+
+  TRANSFER_IN_PROGRESS = 0;
+  TRANSFER_FAILED = 1;
```
Comment (ethan-stride) [inline proto/stride/stakeibc/lsm.proto:13]: If transfer failed, it could be for temporary reasons so we would try again eventually or once it fails to transfer is it done forever?  If the former, what is the trigger for trying to resend all transfer failed tokens?  If we immediately retry then we don't need this state, if we retry at a predictable set time in the future, do we need to track how many times this token has failed to transfer and give up if the error is persistent?

File: proto/stride/stakeibc/validator.proto [hunk 3]
```diff
@@ -16,7 +16,15 @@ message ValidatorExchangeRate {
...
-  string delegation_amt = 5 [
+  string balanced_delegation = 5 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable) = false
+  ];
+  string unbalanced_delegation = 8 [
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable) = false
+  ];
+  string cumulative_lsm_intake = 9 [
```
Comment (ethan-stride) [inline proto/stride/stakeibc/validator.proto:27]: This number only goes up.  But why do we need it?  Over time it could become larger than the total amount of ATOM which exists if people were to loop for some unlikely reason.  This is unrelated to any global limits on amount of LSM possible at any given time.

File: x/stakeibc/keeper/lsm_token_deposit.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+func (k Keeper) SetLSMTokenDeposit(ctx sdk.Context, deposit types.LSMTokenDeposit) {
+	// TODO [LSM]
+}
+
+func (k Keeper) RemoveLSMTokenDeposit(ctx sdk.Context, denom string) {
+	// TODO [LSM]
+}
+
+func (k Keeper) GetLSMTokenDeposit(ctx sdk.Context, denom string) (deposit types.LSMTokenDeposit, found bool) {
```
Comment (ethan-stride) [inline x/stakeibc/keeper/lsm_token_deposit.go:17]: For Remove and Get don't we also need the validator information to arrive at a single specific deposit?  The prefix store will use chain_id then validator then denom as key prefixes, the CTX can tell us the chain_id but can it also tell us the validator or should this also be an input (or return multiple deposits across all validators which match this denom)?

File: x/stakeibc/keeper/lsm_token_deposit.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	// TODO [LSM]
+	return
+}
+
+func (k Keeper) GetAllLSMTokenDeposit(ctx sdk.Context) []types.LSMTokenDeposit {
+	// TODO [LSM]
+	return []types.LSMTokenDeposit{}
+}
+
+func (k Keeper) AddLSMTokenDeposit(ctx sdk.Context, deposit types.LSMDepositStatus) {
```
Comment (ethan-stride) [inline x/stakeibc/keeper/lsm_token_deposit.go]: Debatable if this should be a separate function, why not make this the default behavior for the SetLSMTokenDeposit?  You always want to call Add and never call Set right because a direct call to set in the case when there already is a deposit for this chain+val+denom will overwrite the existing with the second.

File: x/stakeibc/keeper/lsm_token_deposit.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+	// See if a deposit already exists for this denom
+	// If so, increment the amount
+	// otherwise, create a new deposit
+}
+
+func (k Keeper) UpdateLSMTokenDepositStatus(ctx sdk.Context, deposit types.LSMDepositStatus, status types.LSMDepositStatus) {
+	// TODO [LSM]
+}
+
+func (k Keeper) GetLSMDepositsForHostZone(ctx sdk.Context, deposit types.LSMDepositStatus, chainId string) []types.LSMDepositStatus {
```
Comment (ethan-stride) [inline x/stakeibc/keeper/lsm_token_deposit.go]: Do we want a similar version of this function for validator as well to be used when rebalancing?

File: x/stakeibc/keeper/lsm_token_deposit.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+func (k Keeper) AddLSMTokenDeposit(ctx sdk.Context, deposit types.LSMDepositStatus) {
+	// TODO [LSM]
+	// See if a deposit already exists for this denom
+	// If so, increment the amount
+	// otherwise, create a new deposit
+}
+
+func (k Keeper) UpdateLSMTokenDepositStatus(ctx sdk.Context, deposit types.LSMDepositStatus, status types.LSMDepositStatus) {
```
Comment (ethan-stride) [inline x/stakeibc/keeper/lsm_token_deposit.go]: Wrong type on variable deposit for this and 2-3 other methods here.  IDE probably autocompleted LSMDepositStatus but any of the variables named deposit are supposed to be LSMTokenDeposit types.

**Review by ethan-stride — COMMENTED** [review]
Overall makes sense, only actual error I saw was a few wrong types on params for the keeper functions.  Left a few questions on why certain fields or functions are separate as they seem to be repeating other existing fields or functions.

File: proto/stride/stakeibc/lsm.proto (hunk 1 above)
Comment (sampocs) [reply proto/stride/stakeibc/lsm.proto:27]: Very good question. At one point, there was an ICA we needed to send where we needed to specify the record ID in the transaction. It's possible that it's no longer necessary now that we're not hoarding though - I'll double check the spec!

File: proto/stride/stakeibc/lsm.proto (hunk 2 above)
Comment (sampocs) [reply proto/stride/stakeibc/lsm.proto:13]: There's no expected failure scenario that I'm aware of. So a failure would likely indicate that something unpredictable and very wrong happened - at which point we would just resolve it in an upgrade

File: proto/stride/stakeibc/validator.proto (hunk 3 above)
Comment (sampocs) [reply proto/stride/stakeibc/validator.proto:27]: it's only there for the slashing safety mechanism. One other option I considered is we can reset it every time we get a response from the query. I'd be in favor of this approach, but we'd need a good variable name otherwise I could see it causing more confusion

maybe:
`progress_towards_exchange_rate_query`
?

File: proto/stride/stakeibc/lsm.proto (hunk 1 above)
Comment (sampocs) [reply proto/stride/stakeibc/lsm.proto:27]: Ah okay - just checked. It was previously used for the `withdrawRewards` transaction (when we were hoarding LSM tokens), but now that we're not, it's not needed.

Good catch! I'll remove

File: x/stakeibc/keeper/lsm_token_deposit.go (hunk 7 above)
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit.go]: done

File: x/stakeibc/keeper/lsm_token_deposit.go (hunk 6 above)
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit.go]: If I recall correctly, we never have to grab the deposits for a specific validator (but I could be wrong).

I know for sure during rebalance, we want to grab them for the whole host zone.

Let's hold off on adding a validator version for now and we can always add it in in the future if we find that a function would use it

File: x/stakeibc/keeper/lsm_token_deposit.go (hunk 4 above)
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit.go:17]: Good call out! I think we have two options here:

1. Take three parameters, `chainId, validatorAddress, denom`
2. Take an ID field (that represents the concatenation of the fields)

I don't think this function has a clear use case yet since we typically access all the records from a chain. So I don't have a strong opinion. Maybe we start with option 2 and we can adjust later if needed? Wdyt?

Comment (ethan-stride) [reply x/stakeibc/keeper/lsm_token_deposit.go]: As I have continued to think about it more, the denom contains information about the validator (and implicitly the chain_id) already.  You would never be able to have the same denom present on more than one validator.  Handing in a single field like denom would force and iteration over the prefix store whereas all 3 params would allow immediate lookup but this small efficiency different probably doesn't matter.  It is making me wonder if the validator address should be part of the prefix at all (if the prefix store was just chain_id and then denom, you can lookup a specific deposit and get the validator address from the deposit without it being a key in the prefix store)

File: proto/stride/stakeibc/validator.proto (hunk 3 above)
Comment (ethan-stride) [reply proto/stride/stakeibc/validator.proto:27]: Sure sounds reasonable and makes it clearer where it will be used.

File: x/stakeibc/keeper/lsm_token_deposit.go (hunk 4 above)
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit.go]: that's a good point - I think we can remove the validator from the prefix. The only purpose before was when we had to look up all the deposits associated with a given validator, but that's also no longer relevant now that we're not hoarding

Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit.go]: chatted offline - agreed that validator field is not necessary, but we still need chainId for efficient lookups
